### PR TITLE
[RPS-419] Add S3 publish and unpublish scxml action

### PR DESCRIPTION
**Pull request by wojtek-oledzki**
Two simple actions to add or remove public ACL on S3 objects embedded in
the workflow document.

This version does no handle S3 errors yet.

### CODE REVIEW

File: repository-data/application/src/main/resources/hcm-config/configuration/modules/documentworkflow.scxml [hunk 1]
```diff
@@ -356,6 +358,9 @@
 
                 <!-- target-less transition to publish the document when no event payload parameter targetDate is provided -->
                 <transition event="publish" cond="!_event.data?.targetDate">
+                    <!-- Trigger publication of external files -->
+                    <externalstorage:externalFilePublish unpublishedVariant="unpublished"/>
```
Comment (jacsamell) [inline repository-data/application/src/main/resources/hcm-config/configuration/modules/documentworkflow.scxml]: If the variant you need is always unpublished you don't need to specify it as a variable. Just access it directly from the document handle.

File: repository-data/application/src/main/resources/hcm-config/configuration/modules/documentworkflow.scxml [hunk 2]
```diff
@@ -430,6 +435,9 @@
                     <!-- trigger the creation of searchable flag being false on all published datasets on the published version after it's been copied
                          to trigger the published datasets to be updated if applicable -->
                     <nhsdigital:populateSearchableFlag variant="published" depublishing="true"/>
+                    
+		    <!-- Trigger publication of external files -->
+                    <externalstorage:externalFileUnpublish publishedVariant="published"/>
```
Comment (jacsamell) [inline repository-data/application/src/main/resources/hcm-config/configuration/modules/documentworkflow.scxml]: Again would advise not to pass explicit variants

File: repository-data/application/src/main/resources/hcm-config/configuration/modules/scxmlregistry.yaml [hunk 3]
```diff
@@ -2,6 +2,14 @@
 definitions:
   config:
     /hippo:configuration/hippo:modules/scxmlregistry/hippo:moduleconfig/hipposcxml:definitions/documentworkflow:
+      /externalFilePublish:
+        hipposcxml:classname: uk.nhs.digital.externalstorage.workflow.action.ExternalFilePublishAction
```
Comment (jacsamell) [inline repository-data/application/src/main/resources/hcm-config/configuration/modules/scxmlregistry.yaml]: Under advice from @ziemsky we changed the package structure for these. Have a look at mine :slightly_smiling_face:

File: cms/src/main/java/uk/nhs/digital/externalstorage/workflow/action/ExternalFilePublishAction.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    @SuppressWarnings("unused")
+    public void setUnpublishedVariant(String unpublishedVariant) {
+        setParameter("unpublishedVariantExpr", unpublishedVariant);
+    }
+
+    @Override
+    protected ExternalFilePublishTask createWorkflowTask() {
+        return new ExternalFilePublishTask(HippoServiceRegistry.getService(S3Connector.class));
```
Comment (jacsamell) [inline cms/src/main/java/uk/nhs/digital/externalstorage/workflow/action/ExternalFilePublishAction.java]: I'd put the `HippoServiceRegistry.getService(S3Connector.class)` in initTask not in the constructor

File: cms/src/main/java/uk/nhs/digital/externalstorage/workflow/action/ExternalFilePublishAction.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.commons.scxml2.model.ModelException;
+import org.onehippo.cms7.services.HippoServiceRegistry;
+import org.onehippo.repository.documentworkflow.action.AbstractDocumentTaskAction;
+
+import uk.nhs.digital.externalstorage.s3.S3Connector;
+import uk.nhs.digital.externalstorage.workflow.task.ExternalFilePublishTask;
+
+public class ExternalFilePublishAction extends AbstractDocumentTaskAction<ExternalFilePublishTask> {
+
+    private static final long serialVersionUID = 1L;
```
Comment (jacsamell) [inline cms/src/main/java/uk/nhs/digital/externalstorage/workflow/action/ExternalFilePublishAction.java]: Why do you need this?

File: cms/src/main/java/uk/nhs/digital/externalstorage/workflow/action/ExternalFileUnpublishAction.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package uk.nhs.digital.externalstorage.workflow.action;
+
+import org.apache.commons.scxml2.SCXMLExpressionException;
+import org.apache.commons.scxml2.model.ModelException;
+import org.onehippo.cms7.services.HippoServiceRegistry;
+import org.onehippo.repository.documentworkflow.action.AbstractDocumentTaskAction;
+import uk.nhs.digital.externalstorage.s3.S3Connector;
+import uk.nhs.digital.externalstorage.workflow.task.ExternalFileUnpublishTask;
+
+public class ExternalFileUnpublishAction extends AbstractDocumentTaskAction<ExternalFileUnpublishTask> {
```
Comment (jacsamell) [inline cms/src/main/java/uk/nhs/digital/externalstorage/workflow/action/ExternalFileUnpublishAction.java]: Same as other action, don't think you need the parameters for explicit variants and don't pass parameters into the constructor.

File: cms/src/main/java/uk/nhs/digital/externalstorage/workflow/action/ExternalFilePublishAction.java (hunk 5 above)
Comment (wojtek-oledzki) [reply cms/src/main/java/uk/nhs/digital/externalstorage/workflow/action/ExternalFilePublishAction.java]: copy and paste from Minos code - removing.

**Comment by wojtek-oledzki** [comment]
agreed, moving tasks around and removing "variant" configuration.

File: cms/src/main/java/uk/nhs/digital/externalstorage/workflow/AbstractExternalFileTask.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+import java.rmi.RemoteException;
+import javax.jcr.Node;
+import javax.jcr.NodeIterator;
+import javax.jcr.RepositoryException;
+import javax.jcr.query.Query;
+import javax.jcr.query.QueryResult;
+
+public abstract class AbstractExternalFileTask extends AbstractDocumentTask {
+
+    protected DocumentVariant variant;
```
Comment (jacsamell) [inline cms/src/main/java/uk/nhs/digital/externalstorage/workflow/AbstractExternalFileTask.java]: Don't have protected fields, do you even need variant anymore?

File: cms/src/main/java/uk/nhs/digital/externalstorage/modules/S3ConnectorServiceRegistrationModule.java [hunk 8]
```diff
@@ -48,17 +48,15 @@ protected void doShutdown() {
...
 
-        String endpoint = reg.getServiceEndpoint("s3");
         if (!s3Endpoint.isEmpty()) {
-            endpoint = s3Endpoint;
+            s3Builder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(s3Endpoint, s3Region));
         }
-        s3.setEndpoint(endpoint);
 
-        return s3;
+        return s3Builder.build();
```
Comment (jacsamell) [inline cms/src/main/java/uk/nhs/digital/externalstorage/modules/S3ConnectorServiceRegistrationModule.java]: 👍

File: cms/src/main/java/uk/nhs/digital/externalstorage/workflow/externalFilePublish/ExternalFilePublishAction.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    @Override
+    protected ExternalFilePublishTask createWorkflowTask() {
+        return new ExternalFilePublishTask();
+    }
+
+    @Override
+    protected void initTask(ExternalFilePublishTask task) throws ModelException, SCXMLExpressionException {
+        super.initTask(task);
+        task.setS3Connector(HippoServiceRegistry.getService(S3Connector.class));
+        task.setVariantState(HippoStdNodeType.UNPUBLISHED);
```
Comment (jacsamell) [inline cms/src/main/java/uk/nhs/digital/externalstorage/workflow/externalFilePublish/ExternalFilePublishAction.java]: 👍

File: cms/src/main/java/uk/nhs/digital/externalstorage/workflow/AbstractExternalFileTask.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+import javax.jcr.NodeIterator;
+import javax.jcr.RepositoryException;
+import javax.jcr.query.Query;
+import javax.jcr.query.QueryResult;
+
+public abstract class AbstractExternalFileTask extends AbstractDocumentTask {
+
+    private String variantState;
+
+    protected S3Connector s3Connector;
```
Comment (jacsamell) [inline cms/src/main/java/uk/nhs/digital/externalstorage/workflow/AbstractExternalFileTask.java]: Can you make s3 connector private as well? You could pass it into the setTargetState method but if all you do is call HippoServiceRegistry.getService(S3Connector.class) you don't even need to set this on the abstract class really. you could just get it in each task when you need it. I don't mind either way as long as the local variables are private.

File: cms/src/main/java/uk/nhs/digital/externalstorage/workflow/externalFileUnpublish/ExternalFileUnpublishTask.java [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package uk.nhs.digital.externalstorage.workflow.externalFileUnpublish;
+
+import uk.nhs.digital.externalstorage.workflow.AbstractExternalFileTask;
+
+public class ExternalFileUnpublishTask extends AbstractExternalFileTask {
+
+    protected void setTargetStatus(final String resource, final String documentPath) {
```
Comment (jacsamell) [inline cms/src/main/java/uk/nhs/digital/externalstorage/workflow/externalFileUnpublish/ExternalFileUnpublishTask.java]: You don't use documentPath parameter, delete it and add S3Connector instead?

**Pull request merged**